The job **`x86_64-gnu-pre-stabilization`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27544853932/job/81415622633) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/81415622633) [(plain)](https://github.com/rust-lang/rust/commit/ca98ee8124ccb837f91bf94a74c5f050bd10b837/checks/81415622633/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] tests/ui/lint/dropping_copy_types-issue-125189.rs ... ok
test [ui] tests/ui/lint/dropping_copy_types.rs ... ok
test [ui] tests/ui/lint/dropping_references-can-fixed.rs ... ok
test [ui] tests/ui/lint/dropping_copy_types-macros.rs ... ok
test [ui] tests/ui/lint/empty-cfg-predicate.rs#highvers ... FAILED
test [ui] tests/ui/lint/empty-cfg-predicate.rs#novers ... FAILED
test [ui] tests/ui/lint/empty-cfg-predicate.rs#lowvers ... FAILED
test [ui] tests/ui/lint/expansion-time-include.rs ... ignored, ignored used by another main test file ((used by `./expansion-time.rs`))
test [ui] tests/ui/lint/dropping_references.rs ... ok
test [ui] tests/ui/lint/expansion-time.rs ... ok
test [ui] tests/ui/lint/empty-lint-attributes.rs ... ok
test [ui] tests/ui/lint/expect-future_breakage-crash-issue-126521.rs ... ok
---
- error: aborting due to 1 previous error
+ warning: use of empty `cfg(any())`
+   --> $DIR/cfg-empty-any-all.rs:3:7
+    |
+ LL | #[cfg(any())]  // Equivalent to cfg(false)
+    |       ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #[cfg(any())]  // Equivalent to cfg(false)
+ LL + #[cfg(false)]  // Equivalent to cfg(false)
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/cfg-empty-any-all.rs:6:7
+    |
+ LL | #[cfg(all())]  // Equivalent to cfg(true)
+    |       ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+ help: consider using a boolean literal
+    |
+ LL - #[cfg(all())]  // Equivalent to cfg(true)
+ LL + #[cfg(true)]  // Equivalent to cfg(true)
+    |
+ 
+ error: aborting due to 1 previous error; 2 warnings emitted
20 
21 For more information about this error, try `rustc --explain E0425`.
---
-   --> /checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs:6:7
+ warning: use of empty `cfg(any())`
+   --> $DIR/cfg-empty-any-all.rs:3:7
+    |
+ LL | #[cfg(any())]  // Equivalent to cfg(false)
+    |       ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #[cfg(any())]  // Equivalent to cfg(false)
+ LL + #[cfg(false)]  // Equivalent to cfg(false)
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/cfg-empty-any-all.rs:6:7
+    |
+ LL | #[cfg(all())]  // Equivalent to cfg(true)
+    |       ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+ help: consider using a boolean literal
+    |
+ LL - #[cfg(all())]  // Equivalent to cfg(true)
+ LL + #[cfg(true)]  // Equivalent to cfg(true)
+    |
+ 
+ error: aborting due to 1 previous error; 2 warnings emitted


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args conditional-compilation/cfg-empty-any-all.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg-empty-any-all" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0425]: cannot find value `Disabled` in this scope
##[error]  --> /checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs:10:13
   |
LL |     let _ = Disabled; //~ ERROR: cannot find value `Disabled`
   |             ^^^^^^^^ not found in this scope
   |
note: found an item that was configured out
  --> /checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs:4:8
   |
LL | #[cfg(any())]  // Equivalent to cfg(false)
   |          -- the item is gated here
LL | struct Disabled;
   |        ^^^^^^^^
help: consider importing this unit variant
   |
LL + use std::backtrace::BacktraceStatus::Disabled;
   |

warning: use of empty `cfg(any())`
##[warning]  --> /checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs:3:7
   |
LL | #[cfg(any())]  // Equivalent to cfg(false)
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #[cfg(any())]  // Equivalent to cfg(false)
LL + #[cfg(false)]  // Equivalent to cfg(false)
   |

warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/conditional-compilation/cfg-empty-any-all.rs:6:7
   |
LL | #[cfg(all())]  // Equivalent to cfg(true)
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
help: consider using a boolean literal
   |
LL - #[cfg(all())]  // Equivalent to cfg(true)
LL + #[cfg(true)]  // Equivalent to cfg(true)
   |

error: aborting due to 1 previous error; 2 warnings emitted

For more information about this error, try `rustc --explain E0425`.
------------------------------------------

---- [ui] tests/ui/conditional-compilation/cfg-empty-any-all.rs stdout end ----
---- [ui] tests/ui/lint/empty-cfg-predicate.rs#highvers stdout ----
Saved the actual stderr to `/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.highvers/empty-cfg-predicate.highvers.stderr`
diff of stderr:

- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = Some(
-     RustcVersion {
-         major: 1,
-         minor: 89,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
---

30 LL + #[cfg(false)]
31    |
32 
- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = Some(
-     RustcVersion {
-         major: 1,
-         minor: 89,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
---

57 LL + #[cfg(true)]
58    |
59 
- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = Some(
-     RustcVersion {
-         major: 1,
-         minor: 89,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = None
68 error: aborting due to 2 previous errors
69 
70 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/empty-cfg-predicate.rs`

error in revision `highvers`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/empty-cfg-predicate.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--cfg" "highvers" "--check-cfg" "cfg(test,FALSE,novers,lowvers,highvers)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.highvers" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zlint-rust-version=1.89.0"
stdout: none
--- stderr -------------------------------
error: use of empty `cfg(any())`
##[error]  --> /checkout/tests/ui/lint/empty-cfg-predicate.rs:11:7
   |
LL | #[cfg(any())]
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
note: the lint level is defined here
  --> /checkout/tests/ui/lint/empty-cfg-predicate.rs:8:9
   |
LL | #![deny(empty_cfg_predicate)]
   |         ^^^^^^^^^^^^^^^^^^^
---
   |
LL | #[cfg(all())]
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
help: consider using a boolean literal
   |
LL - #[cfg(all())]
LL + #[cfg(true)]
   |

error: aborting due to 2 previous errors
------------------------------------------

---- [ui] tests/ui/lint/empty-cfg-predicate.rs#highvers stdout end ----
---- [ui] tests/ui/lint/empty-cfg-predicate.rs#novers stdout ----
Saved the actual stderr to `/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.novers/empty-cfg-predicate.novers.stderr`
diff of stderr:

- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = None
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
-         minor: 88,
-         patch: 0,
-     },
- )
9 error: use of empty `cfg(any())`
10   --> $DIR/empty-cfg-predicate.rs:11:7
11    |

24 LL + #[cfg(false)]
25    |
26 
- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = None
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
-         minor: 88,
-         patch: 0,
-     },
- )
35 error: use of empty `cfg(all())`
36   --> $DIR/empty-cfg-predicate.rs:15:7
37    |

45 LL + #[cfg(true)]
46    |
47 
- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = None
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = None
50 error: aborting due to 2 previous errors
51 
52 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/empty-cfg-predicate.rs`

error in revision `novers`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/empty-cfg-predicate.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--cfg" "novers" "--check-cfg" "cfg(test,FALSE,novers,lowvers,highvers)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.novers" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: use of empty `cfg(any())`
##[error]  --> /checkout/tests/ui/lint/empty-cfg-predicate.rs:11:7
   |
LL | #[cfg(any())]
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
note: the lint level is defined here
  --> /checkout/tests/ui/lint/empty-cfg-predicate.rs:8:9
   |
LL | #![deny(empty_cfg_predicate)]
   |         ^^^^^^^^^^^^^^^^^^^
---
   |
LL | #[cfg(all())]
   |       ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
help: consider using a boolean literal
   |
LL - #[cfg(all())]
LL + #[cfg(true)]
   |

error: aborting due to 2 previous errors
------------------------------------------

---- [ui] tests/ui/lint/empty-cfg-predicate.rs#novers stdout end ----
---- [ui] tests/ui/lint/empty-cfg-predicate.rs#lowvers stdout ----
Saved the actual stderr to `/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.lowvers/empty-cfg-predicate.lowvers.stderr`
diff of stderr:

- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = Some(
-     RustcVersion {
-         major: 1,
-         minor: 87,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
-         minor: 88,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] self.msrv = Some(
-     RustcVersion {
-         major: 1,
-         minor: 87,
-         patch: 0,
-     },
- )
- [compiler/rustc_errors/src/lib.rs:1298:48] diagnostic.rust_version() = Some(
-     RustcVersion {
-         major: 1,
---
The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/empty-cfg-predicate.rs`

error in revision `lowvers`: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/empty-cfg-predicate.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--cfg" "lowvers" "--check-cfg" "cfg(test,FALSE,novers,lowvers,highvers)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/empty-cfg-predicate.lowvers" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zlint-rust-version=1.87.0"
stdout: none
stderr: none

---- [ui] tests/ui/lint/empty-cfg-predicate.rs#lowvers stdout end ----
---- [ui] tests/ui/lint/unused-features/unused-language-features.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused-features/unused-language-features/unused-language-features.stderr`
diff of stderr:

+ warning: use of empty `cfg(all())`
+   --> $DIR/unused-language-features.rs:15:13
+    |
+ LL | #![cfg_attr(all(), feature(negative_impls))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(negative_impls))]
+ LL + #![cfg_attr(true, feature(negative_impls))]
+    |
+ 
+ warning: use of empty `cfg(any())`
+   --> $DIR/unused-language-features.rs:19:13
+    |
+ LL | #![cfg_attr(any(), feature(never_type))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(any(), feature(never_type))]
+ LL + #![cfg_attr(false, feature(never_type))]
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/unused-language-features.rs:15:13
+    |
+ LL | #![cfg_attr(all(), feature(negative_impls))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(negative_impls))]
+ LL + #![cfg_attr(true, feature(negative_impls))]
+    |
+ 
+ warning: use of empty `cfg(any())`
+   --> $DIR/unused-language-features.rs:19:13
+    |
+ LL | #![cfg_attr(any(), feature(never_type))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(any(), feature(never_type))]
+ LL + #![cfg_attr(false, feature(never_type))]
+    |
+ 
1 error: feature `coroutines` is declared but not used
2   --> $DIR/unused-language-features.rs:5:12
3    |
---
+    |
+ LL | #![cfg_attr(all(), feature(negative_impls))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(negative_impls))]
+ LL + #![cfg_attr(true, feature(negative_impls))]
+    |
+ 
+ warning: use of empty `cfg(any())`
+   --> $DIR/unused-language-features.rs:19:13
+    |
+ LL | #![cfg_attr(any(), feature(never_type))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(any(), feature(never_type))]
+ LL + #![cfg_attr(false, feature(never_type))]
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/unused-language-features.rs:15:13
+    |
+ LL | #![cfg_attr(all(), feature(negative_impls))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(negative_impls))]
+ LL + #![cfg_attr(true, feature(negative_impls))]
+    |
+ 
+ warning: use of empty `cfg(any())`
+   --> $DIR/unused-language-features.rs:19:13
+    |
+ LL | #![cfg_attr(any(), feature(never_type))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(any(), feature(never_type))]
+ LL + #![cfg_attr(false, feature(never_type))]
+    |
+ 
+ error: aborting due to 5 previous errors; 4 warnings emitted


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/unused-features/unused-language-features.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/unused-features/unused-language-features.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused-features/unused-language-features" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-language-features.rs:15:13
   |
LL | #![cfg_attr(all(), feature(negative_impls))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(negative_impls))]
LL + #![cfg_attr(true, feature(negative_impls))]
   |

warning: use of empty `cfg(any())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-language-features.rs:19:13
   |
LL | #![cfg_attr(any(), feature(never_type))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
help: consider using a boolean literal
   |
LL - #![cfg_attr(any(), feature(never_type))]
LL + #![cfg_attr(false, feature(never_type))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-language-features.rs:15:13
   |
LL | #![cfg_attr(all(), feature(negative_impls))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(negative_impls))]
LL + #![cfg_attr(true, feature(negative_impls))]
   |

warning: use of empty `cfg(any())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-language-features.rs:19:13
   |
LL | #![cfg_attr(any(), feature(never_type))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(any(), feature(never_type))]
LL + #![cfg_attr(false, feature(never_type))]
   |

error: feature `coroutines` is declared but not used
##[error]  --> /checkout/tests/ui/lint/unused-features/unused-language-features.rs:5:12
   |
---
+    |
+ LL | #![cfg_attr(all(), feature(slice_ptr_get))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(slice_ptr_get))]
+ LL + #![cfg_attr(true, feature(slice_ptr_get))]
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/unused-library-features.rs:11:13
+    |
+ LL | #![cfg_attr(all(), feature(slice_ptr_get))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(slice_ptr_get))]
+ LL + #![cfg_attr(true, feature(slice_ptr_get))]
+    |
+ 
1 warning: the feature `is_sorted` has been stable since 1.82.0 and no longer requires an attribute to enable
2   --> $DIR/unused-library-features.rs:7:12
3    |
---
+    |
+ LL | #![cfg_attr(all(), feature(slice_ptr_get))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: `#[warn(empty_cfg_predicate)]` on by default
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(slice_ptr_get))]
+ LL + #![cfg_attr(true, feature(slice_ptr_get))]
+    |
+ 
+ warning: use of empty `cfg(all())`
+   --> $DIR/unused-library-features.rs:11:13
+    |
+ LL | #![cfg_attr(all(), feature(slice_ptr_get))]
+    |             ^^^^^
+    |
+    = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
+    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+ help: consider using a boolean literal
+    |
+ LL - #![cfg_attr(all(), feature(slice_ptr_get))]
+ LL + #![cfg_attr(true, feature(slice_ptr_get))]
+    |
+ 
+ error: aborting due to 2 previous errors; 3 warnings emitted


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/unused-features/unused-library-features.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/unused-features/unused-library-features.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused-features/unused-library-features" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-library-features.rs:11:13
   |
LL | #![cfg_attr(all(), feature(slice_ptr_get))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(slice_ptr_get))]
LL + #![cfg_attr(true, feature(slice_ptr_get))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-library-features.rs:11:13
   |
LL | #![cfg_attr(all(), feature(slice_ptr_get))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(slice_ptr_get))]
LL + #![cfg_attr(true, feature(slice_ptr_get))]
   |

warning: the feature `is_sorted` has been stable since 1.82.0 and no longer requires an attribute to enable
##[warning]  --> /checkout/tests/ui/lint/unused-features/unused-library-features.rs:7:12
   |
---
   |
LL | #![cfg_attr(all(), feature(rustc_attrs))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(rustc_attrs))]
LL + #![cfg_attr(true, feature(rustc_attrs))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> $DIR/used-language-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(rustc_attrs))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(rustc_attrs))]
LL + #![cfg_attr(true, feature(rustc_attrs))]
   |

warning: 2 warnings emitted


---
To only update this specific test, also pass `--test-args lint/unused-features/used-language-features.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/unused-features/used-language-features.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused-features/used-language-features" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/used-language-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(rustc_attrs))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(rustc_attrs))]
LL + #![cfg_attr(true, feature(rustc_attrs))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/used-language-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(rustc_attrs))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(rustc_attrs))]
LL + #![cfg_attr(true, feature(rustc_attrs))]
   |

warning: 2 warnings emitted
------------------------------------------

---
   |
LL | #![cfg_attr(all(), feature(allocator_api))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(allocator_api))]
LL + #![cfg_attr(true, feature(allocator_api))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> $DIR/used-library-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(allocator_api))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(allocator_api))]
LL + #![cfg_attr(true, feature(allocator_api))]
   |

warning: 2 warnings emitted


---
To only update this specific test, also pass `--test-args lint/unused-features/used-library-features.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/lint/unused-features/used-library-features.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused-features/used-library-features" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/used-library-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(allocator_api))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: `#[warn(empty_cfg_predicate)]` on by default
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(allocator_api))]
LL + #![cfg_attr(true, feature(allocator_api))]
   |

warning: use of empty `cfg(all())`
##[warning]  --> /checkout/tests/ui/lint/unused-features/used-library-features.rs:8:13
   |
LL | #![cfg_attr(all(), feature(allocator_api))]
   |             ^^^^^
   |
   = note: this used to be a common pattern before `cfg(true)` and `cfg(false)` were added to the language in Rust 1.88
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
help: consider using a boolean literal
   |
LL - #![cfg_attr(all(), feature(allocator_api))]
LL + #![cfg_attr(true, feature(allocator_api))]
   |

warning: 2 warnings emitted
------------------------------------------

---- [ui] tests/ui/lint/unused-features/used-library-features.rs stdout end ----

failures:
    [ui] tests/ui/conditional-compilation/cfg-empty-any-all.rs
    [ui] tests/ui/lint/empty-cfg-predicate.rs#highvers
    [ui] tests/ui/lint/empty-cfg-predicate.rs#novers
    [ui] tests/ui/lint/empty-cfg-predicate.rs#lowvers
    [ui] tests/ui/lint/unused-features/unused-language-features.rs
    [ui] tests/ui/lint/unused-features/unused-library-features.rs
    [ui] tests/ui/lint/unused-features/used-language-features.rs
    [ui] tests/ui/lint/unused-features/used-library-features.rs

```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/pre-stabilization-ci-job.html).

--- a/compiler/rustc_errors/src/diagnostic.rs
+++ b/compiler/rustc_errors/src/diagnostic.rs
@@ -7,6 +7,7 @@ use std::panic;
 use std::path::PathBuf;
 use std::thread::panicking;
 
+use rustc_ast::attr::version::RustcVersion;
 use rustc_data_structures::sync::{DynSend, DynSync};
 use rustc_error_messages::{DiagArgMap, DiagArgName, DiagArgValue, IntoDiagArg};
 use rustc_lint_defs::{Applicability, LintExpectationId};
@@ -175,6 +176,8 @@ pub struct IsLint {
     pub(crate) name: String,
     /// Indicates whether this lint should show up in cargo's future breakage report.
     has_future_breakage: bool,
+    /// Indicates the minimum rust version this lint applies to
+    rust_version: Option<RustcVersion>,
 }
 
 #[derive(Debug, PartialEq, Eq)]
@@ -306,6 +309,11 @@ impl DiagInner {
     /// Indicates whether this diagnostic should show up in cargo's future breakage report.
     pub(crate) fn has_future_breakage(&self) -> bool {
         matches!(self.is_lint, Some(IsLint { has_future_breakage: true, .. }))
+    }
+
+    /// Indicates the minimum rust version this lint applies to.
+    pub(crate) fn rust_version(&self) -> Option<RustcVersion> {
+        self.is_lint.as_ref().and_then(|is| is.rust_version)
     }
 
     pub(crate) fn is_force_warn(&self) -> bool {
@@ -1152,8 +1160,13 @@ impl<'a, G: EmissionGuarantee> Diag<'a, G> {
         self
     } }
 
-    pub fn is_lint(&mut self, name: String, has_future_breakage: bool) -> &mut Self {
-        self.is_lint = Some(IsLint { name, has_future_breakage });
+    pub fn is_lint(
+        &mut self,
+        name: String,
+        has_future_breakage: bool,
+        rust_version: Option<RustcVersion>,
+    ) -> &mut Self {
+        self.is_lint = Some(IsLint { name, has_future_breakage, rust_version });
         self
     }
 

--- a/compiler/rustc_errors/src/lib.rs
+++ b/compiler/rustc_errors/src/lib.rs
@@ -46,6 +46,7 @@ pub use diagnostic_impls::{
 };
 pub use emitter::ColorConfig;
 use emitter::{DynEmitter, Emitter};
+use rustc_ast::attr::version::RustcVersion;
 use rustc_data_structures::AtomicRef;
 use rustc_data_structures::fx::{FxHashSet, FxIndexMap, FxIndexSet};
 use rustc_data_structures::stable_hash::StableHasher;
@@ -365,6 +366,9 @@ struct DiagCtxtInner {
     /// The file where the ICE information is stored. This allows delayed_span_bug backtraces to be
     /// stored along side the main panic backtrace.
     ice_file: Option<PathBuf>,
+
+    /// Controlled by `-Z lint-rust-version`; this allows avoiding emitting lints which would raise MSRV.
+    msrv: Option<RustcVersion>,
 }
 
 /// A key denoting where from a diagnostic was stashed.
@@ -479,6 +483,11 @@ impl DiagCtxt {
         self
     }
 
+    pub fn with_msrv(mut self, msrv: RustcVersion) -> Self {
+        self.inner.get_mut().msrv = Some(msrv);
+        self
+    }
+
     pub fn new(emitter: Box<DynEmitter>) -> Self {
         Self { inner: Lock::new(DiagCtxtInner::new(emitter)) }
     }
@@ -526,6 +535,7 @@ impl DiagCtxt {
             future_breakage_diagnostics,
             fulfilled_expectations,
             ice_file: _,
+            msrv: _,
         } = inner.deref_mut();
 
         // For the `Vec`s and `HashMap`s, we overwrite with an empty container to free the
@@ -1193,6 +1203,7 @@ impl DiagCtxtInner {
             future_breakage_diagnostics: Vec::new(),
             fulfilled_expectations: Default::default(),
             ice_file: None,
+            msrv: None,
         }
     }
 
@@ -1305,6 +1316,12 @@ impl DiagCtxtInner {
                     return None;
                 }
             }
+        }
+
+        if let (Some(msrv), Some(diag_msrv)) = (self.msrv, diagnostic.rust_version())
+            && diag_msrv > msrv
+        {
+            return None;
         }
 
         TRACK_DIAGNOSTIC(diagnostic, &mut |mut diagnostic| {

--- a/compiler/rustc_lint_defs/Cargo.toml
+++ b/compiler/rustc_lint_defs/Cargo.toml
@@ -5,6 +5,7 @@ edition = "2024"
 
 [dependencies]
 # tidy-alphabetical-start
+rustc_ast = { path = "../rustc_ast" }
 rustc_data_structures = { path = "../rustc_data_structures" }
 rustc_error_messages = { path = "../rustc_error_messages" }
 rustc_hir_id = { path = "../rustc_hir_id" }

--- a/compiler/rustc_lint_defs/src/lib.rs
+++ b/compiler/rustc_lint_defs/src/lib.rs
@@ -1,6 +1,7 @@
 use std::borrow::Cow;
 use std::fmt::Display;
 
+use rustc_ast::attr::version::RustcVersion;
 use rustc_data_structures::fx::FxIndexSet;
 use rustc_data_structures::stable_hash::{StableCompare, StableHash, StableHashCtxt, StableHasher};
 use rustc_error_messages::{DiagArgValue, IntoDiagArg};
@@ -305,6 +306,9 @@ pub struct Lint {
 
     /// `true` if this lint is unaffected by `-D warnings`
     pub ignore_deny_warnings: bool,
+
+    /// Used to avoid lints which would affect MSRV
+    pub rust_version: Option<RustcVersion>,
 }
 
 /// Extra information for a future incompatibility lint.
@@ -521,7 +525,40 @@ impl Lint {
             crate_level_only: false,
             eval_always: false,
             ignore_deny_warnings: false,
+            rust_version: None,
         }
+    }
+
+    // FIXME(const-hack): This is used so that `declare_lint` can declare an MSRV statically.
+    // `RustcVersion::parse_str_strict` should ideally be used instead.
+    pub const fn parse_rust_version(version: &str) -> RustcVersion {
+        const fn parse_part(input: &mut &[u8]) -> u16 {
+            let mut val = 0;
+            let mut idx = 0;
+            while idx < input.len() {
+                let v = input[idx];
+                match v {
+                    b'0'..=b'9' => {
+                        val = val * 10 + (v - b'0') as u16;
+                    }
+                    b'.' => {
+                        idx += 1;
+                        break;
+                    }
+                    _ => panic!("invalid character in version"),
+                }
+                idx += 1;
+            }
+            *input = input.split_at(idx).1;
+            val
+        }
+
+        let mut bytes = version.as_bytes();
+        let major = parse_part(&mut bytes);
+        let minor = parse_part(&mut bytes);
+        let patch = parse_part(&mut bytes);
+        assert!(bytes.is_empty());
+        RustcVersion { major, minor, patch }
     }
 
     /// Gets the lint's name, with ASCII letters converted to lowercase.
@@ -671,6 +708,7 @@ macro_rules! declare_lint {
         $($field:ident : $val:expr),* $(,)*
      }; )?
      $(@edition $lint_edition:ident => $edition_level:ident;)?
+     $(@msrv = $msrv:literal;)?
      $($v:ident),*) => (
         $(#[$attr])*
         $vis static $NAME: &$crate::Lint = &$crate::Lint {
@@ -687,6 +725,7 @@ macro_rules! declare_lint {
             }),)?
             $(edition_lint_opts: Some(($crate::Edition::$lint_edition, $crate::$edition_level)),)?
             $(eval_always: $eval_always,)?
+            $(rust_version: Some($crate::Lint::parse_rust_version($msrv)),)?
             ..$crate::Lint::default_fields_for_macro()
         };
     );

--- a/compiler/rustc_middle/src/lint.rs
+++ b/compiler/rustc_middle/src/lint.rs
@@ -514,7 +514,7 @@ pub fn emit_lint_base<'a, D: Diagnostic<'a, ()> + 'a>(
             err.disable_suggestions();
         }
 
-        err.is_lint(lint.name_lower(), has_future_breakage);
+        err.is_lint(lint.name_lower(), has_future_breakage, lint.rust_version);
         // Lint diagnostics that are covered by the expect level will not be emitted outside
         // the compiler. It is therefore not necessary to add any information for the user.
         // This will therefore directly call the decorate function which will in turn emit

--- a/compiler/rustc_session/src/errors.rs
+++ b/compiler/rustc_session/src/errors.rs
@@ -82,7 +82,11 @@ pub fn feature_warn_issue(
     // Decorate this as a future-incompatibility lint as in rustc_middle::lint::lint_level
     let lint = UNSTABLE_SYNTAX_PRE_EXPANSION;
     let future_incompatible = lint.future_incompatible.as_ref().unwrap();
-    err.is_lint(lint.name_lower(), /* has_future_breakage */ false);
+    err.is_lint(
+        lint.name_lower(),
+        /* has_future_breakage */ false,
+        /* rust_version */ None,
+    );
     err.warn(lint.desc);
     err.note(format!("for more information, see {}", future_incompatible.reason.reference()));
 

--- a/compiler/rustc_session/src/session.rs
+++ b/compiler/rustc_session/src/session.rs
@@ -1040,6 +1040,10 @@ pub fn build_session(
         dcx = dcx.with_ice_file(ice_file);
     }
 
+    if let Some(msrv) = sopts.unstable_opts.lint_rust_version {
+        dcx = dcx.with_msrv(msrv);
+    }
+
     let host_triple = TargetTuple::from_tuple(config::host_tuple());
     let (host, target_warnings) =
         Target::search(&host_triple, sopts.sysroot.path(), sopts.unstable_opts.unstable_options)

